src/systems/filecoin_vm/interpreter/vm_interpreter.go

```diff
@@ -24,162 +25,252 @@ const (
 		}
+
+		deductAmount := minerPenaltyTotal
+		if deductAmount > outTree.Balance(minerOwner) {
+			deductAmount = outTree.Balance(minerOwner)
+		}
+		outTree = _withTransferFundsAssert(outTree, minerOwner, addr.BurntFundsActorAddr, deductAmount)
 	}
 
 	// Invoke cron tick, attributing it to the miner of the first block.
-	cronSender := msgs.Blocks()[0].Miner()
-	cron := _makeCronTickMessage(outTree, cronSender)
-	outTree, r = vmi.ApplyMessage(outTree, cron, cronSender)
-	if r.ExitCode() != exitcode.OK() {
-		panic("cron tick failed")
-	}
+
+	TODO()
+	// TODO: miners shouldn't be able to trigger cron by sending messages;
+	// use ControlActor instead (https://github.com/filecoin-project/specs/issues/665)
+
+	firstMiner := msgs.Blocks()[0].Miner()
+	cronMessage := _makeCronTickMessage(outTree, firstMiner)
+	outTree = _applyMessageBuiltinAssert(outTree, cronMessage, firstMiner)
 
 	return
 }
 
-func (vmi *VMInterpreter_I) ApplyMessage(inTree st.StateTree, message msg.UnsignedMessage, minerAddr addr.Address) (
-	st.StateTree, vmr.MessageReceipt) {
+func (vmi *VMInterpreter_I) ApplyMessage(
+	inTree st.StateTree, message msg.UnsignedMessage, onChainMessageSize int, minerAddr addr.Address) (
+	retTree st.StateTree, retReceipt vmr.MessageReceipt, retMinerPenalty actor.TokenAmount) {
+
+	vmiGasRemaining := message.GasLimit()
 
-	compTree := inTree
-	var outTree st.StateTree
-	var toActor actor.ActorState
-	var err error
+	_applyReturn := func(
+		tree st.StateTree, invocOutput vmr.InvocOutput, exitCode exitcode.ExitCode,
+		gasUsedFrom *addr.Address, gasUsedTo *addr.Address, minerPenaltySpec MinerPenaltySpec) {
 
-	fromActor := compTree.GetActorState(message.From())
+		vmiGasUsed := message.GasLimit().Subtract(vmiGasRemaining)
+		vmiGasUsedFIL := _gasToFIL(vmiGasUsed, message.GasPrice())
+		if gasUsedFrom != nil {
+			tree = _withTransferFundsAssert(tree, *gasUsedFrom, *gasUsedTo, vmiGasUsedFIL)
+		}
+
+		switch minerPenaltySpec {
+		case MinerPenaltySpec_None:
+			retMinerPenalty = actor.TokenAmount(0)
+		case MinerPenaltySpec_ChargeUsedGas:
+			Assert(gasUsedFrom == nil)
+			retMinerPenalty = vmiGasUsedFIL
+		default:
+			Assert(false)
+		}
+
+		retTree = tree
+		retReceipt = vmr.MessageReceipt_Make(invocOutput, exitCode, vmiGasUsed)
+	}
+
+	_applyError := func(
+		tree st.StateTree, errExitCode exitcode.SystemErrorCode,
+		gasUsedFrom *addr.Address, gasUsedTo *addr.Address, minerPenaltySpec MinerPenaltySpec) {
+
+		_applyReturn(
+			tree, vmr.InvocOutput_Make(nil), exitcode.SystemError(errExitCode),
+			gasUsedFrom, gasUsedTo, minerPenaltySpec)
+	}
+
+	_vmiUseGas := func(amount msg.GasAmount) (vmiUseGasOK bool) {
+		vmiGasRemaining, vmiUseGasOK = vmiGasRemaining.SubtractWhileNonnegative(amount)
+		return
+	}
+
+	ok := _vmiUseGas(gascost.OnChainMessage(onChainMessageSize))
+	if !ok {
+		// Invalid message; insufficient gas limit to pay for the on-chain message size.
+		_applyError(inTree, exitcode.OutOfGas, nil, nil, MinerPenaltySpec_ChargeUsedGas)
```

Comment:
This will result in zero penalty because used gas is zero (b/c SubtractWhileNonNegative failed). The miner needs to pay more than the gas limit here: the actual message size cost.